Add additional symbol kinds for `current_function`

rileyshahar opened the pull request:
Resolves #48.

This is a larger list of symbol kinds than the original issue, but I think it's justified. These kinds all tend to be top-level, which means you wouldn't lose other information by including these.

rileyshahar committed:
Add more symbol kinds to current_function

There's little reason not to want to display enums, structs, and
interfaces. They tend to be top-level in languages which define them, so
they aren't competing with anything, like some of the smaller symbol
kinds are.

The updated formatting of the conditions is to keep the line length low.

rileyshahar committed:
Add `module` and `namespace` kinds

These are also generally top-level, and (based on admittedly limited testing)
don't interfere with other kinds.

rileyshahar committed:
Mention included function kinds in docs

This will help people set `kind_labels` by giving them an exhaustive
list of kinds they need to set labels for.

wbthomason commented on lua/lsp-status/current_function.lua:
```diff
@@ -18,7 +18,14 @@ local function current_function_callback(_, _, result, _, _)
   local function_symbols = util.filter(util.extract_symbols(result),
     function(_, v)
-      return v.kind == 'Class' or v.kind == 'Function' or v.kind == 'Method'
+      return (v.kind == 'Class' or
```
Style nitpick: I would prefer that we define a module-local table mapping these names to `true` (e.g. `local scopes = {Function = true, Method = true, ...}`), and then here check if `v.kind` is contained, e.g. by `return scopes[v.kind]`. This makes it easier to update the set of scopes at a later date, and makes for cleaner code at only a small efficiency hit.

rileyshahar replied:
Implemented this! (In the messiest commit history of my life because I thought this was trivial and I could commit directly to the PR branch, but oh well. Presumably gonna get squashed, so *shrugs*.)

Thanks for the suggestion, definitely agree with it. The modularity will also make it easier to make configurable.

wbthomason commented on doc/lsp-status.txt:
```diff
@@ -103,7 +103,8 @@ config({config})                                   *lsp-status.config()*
                   accepted as the symbol currently containing the cursor.• 
                 • `current_function`: Boolean, `true` if the current function
```
What do you think of changing this (here, in the API, and in the other docs) to be called `current_scope` or similar? The old name seems no longer accurate.

rileyshahar replied:
This makes sense to me, but it's also a breaking change. Not sure whether the readability is more important.

wbthomason replied:
Perhaps we merge this as-is, then announce a breaking change and make it in a separate PR?

rileyshahar replied:
Makes sense. The other solution is to keep `current_function` as a (maybe undocumented) alias of `current_scope` so we don't break old configs - not sure how many people keep up-to-date with versioning their neovim plugins.

rileyshahar replied:
But I would merge this, we can discuss naming in another PR - it seems like an adjacent but not identical issue that needs some further discussion.

wbthomason replied:
I agree.

wbthomason reviewed (changes requested):
Thanks, this looks good! I have one naming request for consideration, and one requested style change, but otherwise I think this is good to go.

rileyshahar committed:
Extract valid kinds to a module-level array

As suggested by the initial PR review, this significantly improves
readability and maintainability.

rileyshahar committed:
Fix invalid table syntax

Not sure why I thought table keys could be strings.

rileyshahar committed:
Correctly check `v.kind`

(Why you don't commit to a PR branch even for minor things.)

Shatur95 commented:
Is it possible to use default `require('vim.lsp.protocol').CompletionItemKind`? It is not very convenient to set this setting twice.

wbthomason commented:
@Shatur95 Won't that cause issues if users customize `CompletionItemKind` (e.g. via https://github.com/onsails/lspkind-nvim)?

Also, right now, this is nothing extra to configure. Are you referring to customizing how the current symbol is displayed, or filtering the set of symbol kinds to display?

Shatur95 commented:
> @Shatur95 Won't that cause issues if users customize CompletionItemKind (e.g. via https://github.com/onsails/lspkind-nvim)?

No, this plugins just sets `CompletionItemKind` in an ordinary way, so should work ^)

> Also, right now, this is nothing extra to configure. Are you referring to customizing how the current symbol is displayed, or filtering the set of symbol kinds to display?

I was referring to the display settings. I need to set `CompletionItemKind` and` kind_labels` separately. But, probably, this is okay, because by default, `CompletionItemKind` contains just text without icons.

wbthomason commented:
Ah - I think this is actually a different issue, that of unifying `kind_labels` and `CompletionItemKind`. I would definitely be OK with using `CompletionItemKind` as a default for `kind_labels`, but I think that should happen in a separate PR - this PR is just about expanding the set of kinds which may be displayed, not changing how they're displayed.

nihilistkitten commented:
(+1 to using `CompletionItemKind` as a default - I think `kind_labels` should still be configurable, because you might want different labels on your statusline vs for completion items - but I agree that this is completely orthogonal to this PR.)

Shatur95 commented:
Yes, I just asked, sorry for confusing :)

wbthomason commented:
I think this is good to go. Thanks, @nihilistkitten!
